src/qcodes/dataset/data_set.py

```diff
@@ -1243,14 +1244,22 @@ def _enqueue_results(
         self._raise_if_not_writable()
         interdeps = self._rundescriber.interdeps
 
-        toplevel_params = set(interdeps.dependencies).intersection(set(result_dict))
+        result_parameters = set(result_dict.keys())
+        unused_results = result_parameters.copy()
+
+        toplevel_params = interdeps.top_level_params.intersection(result_parameters)
 
-        new_results: dict[str, dict[str, numpy.ndarray]] = {}
+        new_results: dict[str, dict[str, npt.NDArray]] = {}
 
         for toplevel_param in toplevel_params:
-            inff_params = set(interdeps.inferences.get(toplevel_param, ()))
-            deps_params = set(interdeps.dependencies.get(toplevel_param, ()))
-            all_params = inff_params.union(deps_params).union({toplevel_param})
+            # Transitively collect all parameters that are related to any parameter
+            # in the current tree, including parameters that dependencies are inferred from
+            all_params = interdeps.find_all_parameters_in_tree(toplevel_param)
+            # Only include parameters that are present in result_dict
+            # warn here if missing parameters
```

Comment:
the warning happens later, right?
```suggestion
            # Warning will be issued later if missing parameters
```